test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -47,38 +64,359 @@ class test_memtype_cache : public ucs::test_with_param<ucs_memory_type_t> {
+        mem_buffer::supported_mem_types();
+    const size_t size = 1000000;
+    void *buf2_ptr, *cur_ptr;
+    std::vector<void*> initial_ptrs;
+
+    initial_ptrs.push_back(NULL);
+    initial_ptrs.push_back((void*)ucs_get_page_size());
+
+    for (std::vector<void*>::const_iterator initial_ptr = initial_ptrs.begin();
+         initial_ptr != initial_ptrs.end(); ++initial_ptr) {
+        for (std::vector<ucs_memory_type_t>::const_iterator iter =
+                 supported_mem_types.begin();
+             iter != supported_mem_types.end(); ++iter) {
+
+            std::vector<std::pair<void*, size_t> > released_ptrs;
+
+            /* Create two buffers that possibly will share one page
+             *
+             *                         < shared page >
+             *                            ||    ||
+             *                            \/    ||
+             *        +----------------------+  ||
+             * buf1:  |    |    |    |    |  |  \/
+             *        +----------------------+----------------------+
+             *                        buf2:  |  |    |    |    |    |
+             *                               +----------------------+
+             */
+            cur_ptr          = get_next_ptr(*initial_ptr, 0);
+            mem_buffer *buf1 = allocate_mem_buffer(cur_ptr, size, GetParam());
+
+            cur_ptr          = get_next_ptr(cur_ptr, size);            
+            mem_buffer *buf2 = allocate_mem_buffer(size, *iter);
+
+            test_region_found(*buf1, size);
+            test_region_found(*buf2, size);
+
+            buf2_ptr = buf2->ptr();
+            release_mem_buffer(buf2, &released_ptrs);
+
+            /* buffer `buf2` is released, but need to consider
+             * a shared page with buffer `buf1` */
+            test_ptr_released(buf2_ptr, size, 1);
+
+            test_region_found(*buf1, size);
+
+            release_mem_buffer(buf1, &released_ptrs);
+
+            /* buffer `buf1` and `buf2` are released */
+            test_ptrs_released(&released_ptrs);
+        }
+    }
+}
+
+UCS_TEST_P(test_memtype_cache, different_mem_types) {
+    std::vector<void*> initial_ptrs;
+    std::vector<bool> insert_type;
+
+    initial_ptrs.push_back(NULL);
+    initial_ptrs.push_back((void*)ucs_get_page_size());
+
+    insert_type.push_back(false);
+    insert_type.push_back(true);
+
+    const std::vector<ucs_memory_type_t> supported_mem_types =
+        mem_buffer::supported_mem_types();
+
+    /* The tests try to allocate/simulate two buffers with different memory types */
+    for (std::vector<void*>::const_iterator initial_ptr = initial_ptrs.begin();
+         initial_ptr != initial_ptrs.end(); ++initial_ptr) {
+        for (std::vector<ucs_memory_type_t>::const_iterator iter =
+                 supported_mem_types.begin();
+             iter != supported_mem_types.end(); ++iter) {
+            /* 1. Allocate the same amount of memory for buffers */
+            {
```

Comment:
why do we need to have separate cases with immediate buffer release?

Reply:
this is the simplest test case that can test that after releasing 2nd buffer, we still can access the 1st one